### ISSUE TITLE
Combine clustering and correlation analytics

### ISSUE THREAD

**Issue by bentsherman** [SystemsGenetics/KINC#37]
Based on the discussion from #34, it looks like we may need to create an analytic which can perform both clustering and correlation at once. This is because the correlation analytics can throw away output data which does not meet a threshold (currently 0.5 by default), but the clustering analytics can't do this because they don't know the correlation. As a result, the CMX and CCM files produced from an EMX will be mismatched; the CMX will not contain data for every cluster in the CCM, and there will be no way to match the correlation data with the cluster data. So we have two options:
1. Change the correlation analytics to save all correlations
2. Create an analytic to perform both clustering and correlation

Since the whole reason for thresholding the correlations was to reduce file size, I feel that we have no choice but to take option 2. I'm imaging an analytic with the following arguments:
- input EMX file
- output CCM file
- output CMX file
- clustering method (none, GMM, k-means)
- correlation method (pearson, spearman)
- min samples
- min expression threshold
- min clusters
- max clusters
- criterion (BIC, ICL)
- remove pre-outliers
- remove post-outliers
- min correlation threshold
- max correlation threshold

So essentially an analytic which parallels KINCv1 similarity.

Any thoughts @spficklin @4ctrl-alt-del ?

**Comment by bentsherman** [comment]
I made a Similarity analytic on the `similarity-analytic` branch. It just supports serial for now since the whole serial / MPI / OpenCL interface is about to change drastically, but as a proof of concept I think it works. It takes each gene pair, performs clustering and correlation, and saves any relevant gene pair data to the CCM and CMX. One benefit is that the correlation code doesn't have to do any file I/O; the clustering code just passes along what it loaded.

**Comment by 4ctrl-alt-del** [comment]
Sorry for not replying earlier to this I have been really focused on updating ACE.

While I agree we should not do #1 I strongly disagree that we should do #2. I have a #3...

3. Make very minor changes to CMX data type so it can save specific clusters.. so you can have a gene pair of say cluster 2 and 3 without 0 or 1.

The reason I strongly disagree with #2 is it is committing one of my personal sins of programming design: The natural tendency to conglomerate everything into one massive "do everything" class/function.

If we do #3 it is an easy fix instead of having to make a new giant glob analytic in one giant source file that is really long and hard to read because it does everything.

**Comment by bentsherman** [comment]
Well the Similarity analytic itself doesn't contain any actual clustering / correlation code -- that code has been abstracted into separate classes. If you look at my code on the `similarity-analytic` branch, I think you might find that it isn't doing what you think. In fact, I would argue that the Similarity analytic actually _reduces_ the total amount of code. Think about all of the duplicated code between kmeans, gmm, pearson, and spearman -- they all have some ~500 lines of common boilerplate. The Similarity analytic actually grants us two advantages:
1. Eliminates all of that duplicated code
2. Provides an easy way to incorporate other clustering / correlation methods in the future

Take a look at `similarity.cpp` (and the `GenePair::...` classes that it uses) and let me know what you think.

**Comment by 4ctrl-alt-del** [comment]
Ok it sounds like I am misunderstanding what you are saying. Since reducing the number of lines is a good thing. :)

I'll take a look at similarity when I can but it sounds like your second choice is the way to go so it's fine. Sorry for misunderstanding.

**Comment by 4ctrl-alt-del** [comment]
So after having some more time to think about this I still don't like the direction this is taking KINC from a design and user perspective. You're correct the code base is more compact by doing the similarity but that's not the issue I have.

The genesis of ACE and the whole data/analytic paradigm came as a result of me trying to separate the clustering and correlation algorithms. A main complaint Stephen had about KINC v1 was the fact that if you wanted to change anything you would have to run everything again; for example if you wanted to do a different correlation you would have to re-run the clustering all over again which took ages. Me and Stephen had many long discussions about this and one of the main things he likes about the new KINC is the separation of analytics into as many smaller algorithms as possible.

Making a similarity analytic that basically reinvents KINC v1 that me and Stephen have been trying to move away from feels like moving backwards to me. Also this is a major redesign of KINC and even I wouldn't do something like this without getting Stephen's explicit permission first.

So basically the direction me and Stephen agreed on for where KINC should go is making analytics as atomic and small as possible and the similarity analytic is a complete 180 from that general design.

You out there @spficklin ? :)

**Comment by bentsherman** [comment]
Yeah, I mean, I agree with you on those points, I don't really like that clustering and correlation were coupled either. And modifying the CMX format like you said would eliminate the problem. But another, more practical problem remains, which is that if the clustering analytics aren't able to remove cluster data that corresponds to insignificant correlations, then as we scale to larger datasets the resulting CCMs are going to be quite large. Running GMM on Yeast (~5 MB) produces an ~8 GB CCM, so something like GTEX R1 (~800 MB) would produce... I don't know how many TB. So if we stick with keeping the analytics separated, we'll have to be willing to deal with that kind of file size.

(As a side note, I suppose the extra benefit of eliminating repeated code could still be achieved by having a "clustering" analytic and "correlation" analytic, but I'll leave that for some other time to work on.)

**Comment by 4ctrl-alt-del** [comment]
Practical problems must take precedence, and thanks for pointing that out I did not realize that was such a problem. In that case, I must agree with your idea of merging clustering and correlation back together again because I see no other option. We will of course try to keep the spirit of atomic analytics as much as possible but in this case I see no practical way to do that as you point out.

The best laid plans of mice and men... :)

**Comment by bentsherman** [comment]
Finally managed to combine everything into the Similarity analytic, including OpenCL and MPI. It's pretty ugly in some places, and command-line usage is slightly more painful. But on the bright side, all clustering / correlation methods are now MPI capable! Once RMT is refactored to be cluster-aware, KINC will have full KINCv1 functionality and we could actually start playing around with it then, but I think I would like to write some unit tests first.
